### ISSUE TITLE
Support for mix-in in Mirah

### ISSUE THREAD

**Issue by altamic** [mirah/mirah#234]
TL;DR: I propose an `include` macro that manipulates the AST of the class passed as argument to be mixed-in into the receiver class.

In Ruby a module is a collection of constants and methods. A method in a module may be an instance or a module method:

``` ruby
module M
  C = 'const'
  def inst_method
    'I am an instance method'
  end
  def self.module_method
    'I am a module method'
  end  
end
```

Constants and instance methods appear in a class when the module is included, module methods do not:

``` ruby
class A
  include M
end

A::C  # => "const"
A.new.inst_method # => "I am an instance method"
A.module_method # => NoMethodError: undefined method `module_method' for A:Class
```

Conversely, module methods may be called without creating an encapsulating object, while instance methods may not.

``` ruby
M.module_method # => "I am a module method"
M.inst_method # => NoMethodError: undefined method `inst_method' for M:Module
```

In practice Ruby modules are used for 2 reasons:
1. namespace a group of related constants and/or functions by creating module methods.
2. encapsulate common methods to be mixed-in across multiple classes by defining instance methods.

It is common in Java to define a class with no public constructor, only exposing static constants and/or methods. In other words, namespacing (1) poses no problems in Java. Some "modules" from the Java standard platform are: [Math](http://docs.oracle.com/javase/6/docs/api/java/lang/Math.html), [Arrays](http://docs.oracle.com/javase/7/docs/api/java/util/Arrays.html).

Indeed, Mirah can already do:

```
$ mirah -e 'import java.util.Arrays; a = [1,2,3,4,5].toArray; index = Arrays.binarySearch(a, 4); puts index'
3
```

The problem arises for mix-in (2): in Java there is no easy way to "include" public instance methods and constants of other classes. After some thought, I believe that we could implement mix-in at compile time with a macro and some AST manipulation.

Let's see how with an example. Consider a couple of classes implementing a pattern that we want to share through mix-in:

``` ruby
package mirah.lang

class Singleton
  def self.initialize
    @@instance = Singleton(nil)
  end

  def self.instance
    @@instance || Singleton.new
  end

  private
  def initialize
  end
end

class Comparable
  def <(other)
    (self <=> other) < 0
  end

  def <=(other)
    (self <=> other) <= 0
  end

  def ==(other)
    (self <=> other) == 0
  end

  def >(other)
    (self <=> other) > 0
  end

  def >=(other)
    (self <=> other) >= 0
  end

  def between?(min, max)
    return false if self < min
    return false if self > max
    true
  end
end
```

Suppose to have a macro called `include` extending java.lang.Object that accepts a constant as argument. Let's see a use case where we need to sort some entities, one kind of them being a singleton.

``` ruby
package com.corporation.offices

import mirah.lang.Singleton
import mirah.lang.Comparable

interface IAddress do
  def street:String; end
  def city:String; end
  def country:String; end
end

class Address
  implements IAddress
  include Comparable

  def initialize(street = '', city = 'Rome', country = 'Italy')
    @street  = street
    @city    = city
    @country = country
  end

  attr_reader street:String, city:String, country:String

  def <=>(other:IAddress)
    [(self.country <=> other.country),
     (self.city <=> other.city),
     (self.street <=> other.street)].find {|x| x != 0} || 0
  end
end

class HeadOfficeAddress
  implements IAddress
  include Singleton
  include Comparable

  attr_accessor street:String, city:String, country:String

  def <=>(other:IAddress)
    -1
  end
end
```

when the Mirah parser finds the macro `include` inside the class `Address` expands the AST definition of the `Comparable` argument, that is expected to be a ClassDefinition; the typer will infer the type of a formal parameter
for a given mixed method by examining its implementation based on the information of the container class.
For example, the formal parameter `other` of the instance method `<` that has been mixed in from `Comparable` within `Address` depends on the type of `Address`'s `<=>` instance method. Since the `Address` class has declared that the `<=>` instance method receives an argument of IAddress type, such is the type of the `<` argument. The reasoning is analog for the `<=`, `==`, `>` and `>=` operators. 
Once the type of a method of the mix-in class has been resolved it is permanently added to the container class, so that a method like `between?`, which type depends on methods present in the mix-in, can be correctly resolved by inferring types from the implementation referred to the container class.
The obtained AST would be then equivalent to the following source code:

``` ruby
class Address
  implements IAddress

  def initialize(street = '', city = 'Rome', country = 'Italy')
    @street  = street
    @city    = city
    @country = country
  end

  attr_reader street:String, city:String, country:String

  def <=>(other:IAddress)
    [(self.country <=> other.country),
     (self.city <=> other.city),
     (self.street <=> other.street)].find {|x| x != 0} || 0
  end

  def <(other:IAddress)
    (self <=> other) < 0
  end

  def <=(other:IAddress)
    (self <=> other) <= 0
  end

  def ==(other:IAddress)
    (self <=> other) == 0
  end

  def >(other:IAddress)
    (self <=> other) > 0
  end

  def >=(other:IAddress)
    (self <=> other) >= 0
  end

  def between?(min:IAddress, max:IAddress)
    return false if self < min
    return false if self > max
    true
  end
end
```

The process is then repeated with the `HeadOfficeAddress` class. Now the `Singleton` is mixed in and after doing that, Mirah will substitute in the AST any self reference of the mix-in class `Singleton` with the container that is `HeadOfficeAddress`, merging the result. Note that in general method accessibility should be reset to public before a mix-in is inserted to prevent side effects from previous inclusion(s). Morever, mixed-in constructors should be merged into the receiver's ones. Here it is the code for `HeadOfficeAddress`:

``` ruby
class HeadOfficeAddress
  implements IAddress

  attr_accessor street:String, city:String, country:String

  def <=>(other:IAddress)
    -1
  end

  #include Singleton
  public
  def self.initialize
    @@instance = HeadOfficeAddress(nil)
  end

  def instance:HeadOfficeAddress
    @@instance || HeadOfficeAddress.new
  end

  private
  def initialize
  end

  #include Comparable
  public
  def <(other:IAddress)
    (self <=> other) < 0
  end

  def <=(other:IAddress)
    (self <=> other) <= 0
  end

  def ==(other:IAddress)
    (self <=> other) == 0
  end

  def >(other:IAddress)
    (self <=> other) > 0
  end

  def >=(other:IAddress)
    (self <=> other) >= 0
  end

  def between?(min:IAddress, max:IAddress)
    return false if self < min
    return false if self > max
    true
  end
end
```

Note that the proposed kind of mix-in slightly differs from the Ruby's one. Here `include` adds module methods as well as instance methods. But by doing  so we can having a working mix-in in one move, deferring the implementation of an `extend` macro to augment the receiver class with class methods (in this case with `SingletonClassMethods`).

What do you think about such possibility? Is it technically feasible? I am eager to hear from you some feedback, thank you.
